system-addon/lib/ManualMigration.jsm

```diff
@@ -0,0 +1,93 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {utils: Cu} = Components;
+const {actionCreators: ac, actionTypes: at} = Cu.import("resource://activity-stream/common/Actions.jsm", {});
+const MIGRATION_ENDED_EVENT = "Migration:Ended";
+
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/Services.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this, "MigrationUtils", "resource:///modules/MigrationUtils.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "Prefs", "resource://activity-stream/lib/ActivityStreamPrefs.jsm");
+
+this.ManualMigration = class ManualMigration {
+  constructor() {
+    Services.obs.addObserver(this, MIGRATION_ENDED_EVENT);
+    this._prefs = new Prefs();
+  }
+
+  uninit() {
+    Services.obs.removeObserver(this, MIGRATION_ENDED_EVENT);
+  }
+
+  isMigrationMessageExpired() {
+    let migrationLastShownDate = new Date(this._prefs.get("migrationLastShownDate") * 1000);
+    let today = new Date();
+    // Round down to midnight.
+    today = new Date(today.getFullYear(), today.getMonth(), today.getDate());
+
+    if (migrationLastShownDate < today) {
+      let migrationRemainingDays = this._prefs.get("migrationRemainingDays") - 1;
+
+      this._prefs.set("migrationRemainingDays", migrationRemainingDays);
+      // .valueOf returns a value that is too large to store so we need to divide by 1000.
+      this._prefs.set("migrationLastShownDate", today.valueOf() / 1000);
+
+      if (migrationRemainingDays <= 0) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * While migrationExpired is false and migration message is still displayed
+   * keep checking until expiration date. Broadcast expiration to store.
+   *
+   * @param migrationExpired - Pref flag that is false for the first 3 active days,
+   *                           time in which we display the migration message to the user.
+   */
+  checkMigrationStatus(migrationExpired) {
+    if (migrationExpired === false) {
+      const isExpired = this.isMigrationMessageExpired();
+
+      if (isExpired === true) {
+        this.setMigrationStatus();
```

Comment:
This method seems unnecessarily verbose and awkwardly named because it changes state when it says it's just "check"ing.
Maybe something like
```js
expireIfNecessary(alreadyExpired) {
  if (!alreadyExpired && this.isMigrationMessageExpired()) {
    this.expireMigration();
  }
}
```